pyexchange/tethfinex.py

```diff
@@ -217,43 +219,50 @@ def place_order(self,
 
         assert(isinstance(is_sell, bool))
         assert(isinstance(pay_token, Address))
         assert(isinstance(pay_amount, Wad))
         assert(isinstance(buy_token, Address))
         assert(isinstance(buy_amount, Wad))
-        assert(isinstance(fee_address, Address))
+        assert(isinstance(ethfinex_address, Address))
+        assert(isinstance(exchange_address, Address))
         assert(isinstance(pair, str))
+        # fee rate
+        # use 0.0025 for 0.25% fee
+        # use 0.01   for 1% fee
+        # use 0.05   for 5% fee
+        fee_rate = 0.0025
 
         expiration = int((datetime.datetime.today() + datetime.timedelta(hours=6)).strftime("%s"))
-        order = ZrxOrder(exchange=self.tethfinex,
+        order = ZrxV2Order(exchange=self.tethfinex,
+                         sender=ethfinex_address,
                          maker=Address(self.tethfinex.web3.eth.defaultAccount),
-                         taker=fee_address,
+                         taker=Address('0x0000000000000000000000000000000000000000'),
                          maker_fee=Wad(0),
                          taker_fee=Wad(0),
-                         pay_token=pay_token,
-                         pay_amount=(pay_amount / buy_amount) * buy_amount if is_sell else pay_amount,
-                         buy_token=buy_token,
-                         buy_amount=buy_amount if is_sell else (buy_amount / pay_amount) * pay_amount,
-                         salt=self.tethfinex.random_salt(),
-                         fee_recipient=fee_address,
+                         pay_asset=Asset.deserialize('0xf47261b0E' + str(pay_token)[2:]),
+                         pay_amount=pay_amount,
+                         buy_asset=Asset.deserialize('0xf47261b0E' + str(buy_token)[2:]),
+                         buy_amount=Wad.from_number(float(buy_amount) - (float(buy_amount)/100 * fee_rate * 100)),
+                         salt=self.tethfinex.generate_salt(),
+                         fee_recipient=ethfinex_address,
                          expiration=expiration,
-                         exchange_contract_address=self.tethfinex.address,
-                         ec_signature_r=None,
-                         ec_signature_s=None,
-                         ec_signature_v=None)
+                         exchange_contract_address=exchange_address,
+                         signature=None)
 
         signed_order = self.tethfinex.sign_order(order)
         data = {
             "type": 'EXCHANGE LIMIT',
             "symbol": f"t{pair}",
-            "amount": str(buy_amount) if is_sell else str(f"-{pay_amount}"),
-            "price": str(pay_amount / buy_amount) if is_sell else str(buy_amount / pay_amount),
+            "amount": str(f"-{round(float(pay_amount),6)}") if is_sell else str(round(float(buy_amount),6)),
+            "price": str(round(float(buy_amount / pay_amount),5)) if is_sell else str(round(float(pay_amount / buy_amount),5)),
```

Comment:
instead using hardcodded precision for amount and price we should dynamically get them by calling self.get_symbols (which goes to https://api.ethfinex.com/v1/symbols_details API) - pyexchange should not deal with this though but the external price calculators as keepers, see Trustless Ethfinex keeper https://github.com/makerdao/market-maker-keeper/blob/master/market_maker_keeper/tethfinex_market_maker_keeper.py

Reply:
I hardcodded the precision because i have read them as max values in ethfinex doc.
If you look in https://api.ethfinex.com/v1/symbols_details you can see that the `price_precision` is 5 for each pair and amount precision is missing into this get.

Reply:
I have noticed in https://api.ethfinex.com/v1/symbols_details that `minimun_order_size` and `maximun_order_size` are wrong, i think. For `ethdai` the real minimun size is 25DAI (checked on Ethfinex trustless) but in symbol details it is 6 (but for daieth). Also i noticed that for some get i can use daieth and for another both `ethdai` and `daieth`. Maybe in https://api.ethfinex.com/v1/symbols_details there are details for the classic ethfinex not for tethfinex.

Reply:
well `DAIETH` is actually `ETHDAI` pair on Trustless Ethfinex, I suggest to use what API returns (that is price_precision":5 and "minimum_order_size":"10.0) as below

`{"pair":"daieth","price_precision":5,"initial_margin":"30.0","minimum_margin":"15.0","maximum_order_size":"20000.0","minimum_order_size":"10.0","expiration":"NA","margin":false}`

@plutoegg are these the right details for pair?

Reply:
Hi @grandizzy,
I used `ETHDAI` instead of `DAIETH` for submitting an order because i wanted to leave the same login that there is in `efx-api-node` (pair exists in Trustless Efx). I tested submitting the same order in both python client and directly in trustless exchange and it worked, but the minimun order size is 25 DAI or 0.1 ETH, not 10.
If you look in https://github.com/ethfinex/efx-api-node, in submitting an order description, it uses `ZRXETH` (existing pair in Trustless Ethfinex).
I noticed that for `ZRXETH`, for instance, the same symbol is in Trustless Ethfinex and in api return but for `ETHDAI`, it is in exchange but the api returns `DAIETH` as you could see.

Reply:
I have just noticed that in https://api.ethfinex.com/v1/symbols_details some values changed, for instance yesterday the `minimun_order_size` was 6, as I wrote in my comment. Today is 10 as you wrote.

Reply:
I have to fix the code for submitting a buy order.

Reply:
Hi @grandizzy,
May be that in the DAI market the logic of the buy/sell is the opposite that in the other market ?
Because I have done a sell ETH/DAI order and a buy OMG/ETH order and in trustless ethfinex order history, the both orders are tagged as `BUY`.

Reply:
Price_precisions should always be 5 for all pairs, however it is best to read that from `https://api.ethfinex.com/v1/symbols_details` in case for any reason it ever changes.

Minimum and maximum order sizes are different on Trustless than on regular API. For Trustless the minimums are higher, and can be read from api.ethfinex.com/trustless/v1/get/config for each token. They also change dynamically, adjusted according to price movements, as you observed above.

Reply:
Via API the order is expected as DAIETH, although we flip this in the UI as it is a better user experience for those who expect to see ETH quoted against DAI.

Reply:
Also right now the logic of submitting an order using `efx-api-node` and this python client is little different, I implemented the new v2 version in it leaving the same old order logic.
In `efx-api-node` for submitting an order I have to choose pair `symbol`, the `amount` that i would like to buy/sell of the first symbol in pair and the pair rate `price`, the client will calculate the amount of the second symbol in `symbol`. Instead with this client I have to submit the value of both pair and it will calculate the rate price base of these values, I'm fixing the buy order, I did not understand properly the difference in `DAI` market and `ETH` market with pair symbol order.

Reply:
Hi plutoegg,
Ah ok, understood. I did not understand properly because i have submitted some sell orders via api without problems using `ETHDAI` in `symbol` inside post data but they appears as `BUY` order on trustless efx UI as you explained above.

Reply:
indeed DAIETH pair on trustless ethfinex is different than the others, can you handle this exception in pyexchange code so we could use same code for any pair we may want to trade? I recall I had to make some workarounds to make it work for 0x DAIETH only

Reply:
I'm thinking to add a check when it retrieves price precision on https://api.ethfinex.com/v1/symbols_details, it catches an error when pairs don't exist, because i'm looking in it and not only DAIETH is different.
Can it be good for you ? @grandizzy

Reply:
that's good, though I also think you will have to set amount and price for `DAIETH` pair differently than for other pairs?

Reply:
In theory it should be ok for any pair, only in trustless ethfinex UI will reflect the pair swithed. I tested submitting `SELL` orders using ETHDAI but in UI they are ETH/DAI `BUY` orders. I have to test to create a `BUY` DAIETH order, i think that it will be a `BUY`ETH/DAI order, same of other orders that i made. If i block the way to create a ETHDAI sell/buy orders checking in symbol details via API it should be fine, all my orders for this and other different pair will be only reflected switched on exchange UI but the logic of orders is the same.

Reply:
@PumpkingWok IMO the order that we place should be shown accordingly in UI, otherwise we may have surprises of orders being filled at wrong prices and so. Also synchronizing trades will be really misleading if we placed a `BUY` but we get a `SELL` from API call

Reply:
@grandizzy,
In fact adding the check on pair as i explained above, it should not happen anymore.
I did some `SELL` orders using `ETHDAI` that reflects eth/dai `BUY` orders on UI. Adding the check i can't do it anymore because in symbol details there is only `DAIETH` pair. I have not tried to submit a `BUY` `DAIETH` order but i think that it will be a `BUY` order identical to the same `BUY` order that i did in these days using `ETHDAI`. plutoegg wrote that API should not accept `ETHDAI` but i did some. Maybe he mean that api does not accept `DAIETH` pair symbol ? If yes then it needs to change the logic for the different pair as you told but it is little complicated because `DAIETH` is not only pair that has logic reversed.

Reply:
Hi @grandizzy,
I can't find your last comment regard fees, I have already implement them at code level.
I'm waiting to push a new commit because i would like to understand properly if i can submit both `DAIETH` or `ETHDAI` order via api.Thanks

Reply:
I am surprised that ETHDAI works... I will need to double verify this for myself. It should not be possible at the moment to submit `tETHDAI` orders via api. Could you post an example request and response?

Reply:
Hi plutoegg,
![Screenshot 2019-04-23 at 20 45 06](https://user-images.githubusercontent.com/2848253/56607692-6fec3280-6609-11e9-9d3d-1b6f1bbabc4e.png)

![Screenshot 2019-04-23 at 20 44 33](https://user-images.githubusercontent.com/2848253/56607438-d1f86800-6608-11e9-9477-dd3d76893976.png)
It is the POST data just before api call, It should be a `ETHDAI` `SELL` order (negative `amount`), it submit an order to buy `30 DAI` selling `0.1 ETH`, with price of `300`
On Trustless Etherfinex Exchange it is a `BUY` order.
![Screenshot 2019-04-23 at 20 45 32](https://user-images.githubusercontent.com/2848253/56607781-9dd17700-6609-11e9-91fc-d76048e26ed3.png)
It is the order on top.
The last two botton orders are identical, I did it two days ago, one using python client submitting always a `SELL` order as i have done now and one using directly Ethfinex UI instead. I noticed now that in both cases it is a `BUY` order, @grandizzy so even using directly the UI if i submit a `SELL` order, it becomes a `BUY` order in history.

Reply:
Hi plutoegg,
Did you test it ?
Thanks

Reply:
Hi @grandizzy @plutoegg,
How is doing ? Any news on test ?
Thanks in advance.